From 472abaea65d302a5967242d6392498cdcc7d03fa Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 17 Aug 2023 10:27:38 +0800
Subject: [PATCH 51/53] [Huawei] mm/sharepool: Increase the spa count before
 sp_map_spa_to_mm()

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

We should prevent the spa from being freed before sp_map_spa_to_mm(), or
we would map an invalid spa to a mm_struct. Then UAF occurs.

BUG: KASAN: use-after-free in sp_area_drop+0x60/0x120
Write of size 4 at addr ffff000158052c2c by task test_perf_sp_ad/284

CPU: 0 PID: 284 Comm: test_perf_sp_ad Tainted: G           OE     5.10.0+ #1
Hardware name: linux,dummy-virt (DT)
Call trace:
 dump_backtrace+0x0/0x30c
 show_stack+0x20/0x30
 dump_stack+0x11c/0x174
 print_address_description.constprop.0+0x2c/0x1fc
 __kasan_report+0xe0/0x140
 kasan_report+0x44/0x6c
 check_memory_region+0xf8/0x1a0
 __kasan_check_write+0x34/0x60
 sp_area_drop+0x60/0x120
 remove_vma+0x84/0xa4
 __do_munmap+0x45c/0x5d0
 __mmap_region+0x260/0xa00
 __do_mmap_mm+0x34c/0x5c4
 sp_mmap+0xc4/0x1e0
 sp_map_spa_to_mm+0x188/0x460
 __mg_sp_alloc_nodemask+0x1c4/0x580
 mg_sp_alloc+0x1c/0x30
 dev_sp_alloc+0x108/0x2d0 [sharepool_dev]
 dev_ioctl+0x80/0x3e0 [sharepool_dev]
 __arm64_sys_ioctl+0xd8/0x11c
 el0_svc_common.constprop.0+0xa8/0x240
 do_el0_svc+0x8c/0xc0
 el0_svc+0x20/0x30
 el0_sync_handler+0xb0/0xb4
 el0_sync+0x160/0x180

Allocated by task 253:
 kasan_save_stack+0x28/0x60
 __kasan_kmalloc.constprop.0+0xa4/0xd0
 kasan_kmalloc+0x10/0x20
 kmem_cache_alloc_trace+0xe0/0x5c4
 sp_area_alloc+0x10c/0x684
 __mg_sp_alloc_nodemask+0x150/0x580
 mg_sp_alloc+0x1c/0x30
 dev_sp_alloc+0x108/0x2d0 [sharepool_dev]
 dev_ioctl+0x80/0x3e0 [sharepool_dev]
 __arm64_sys_ioctl+0xd8/0x11c
 el0_svc_common.constprop.0+0xa8/0x240
 do_el0_svc+0x8c/0xc0
 el0_svc+0x20/0x30
 el0_sync_handler+0xb0/0xb4
 el0_sync+0x160/0x180

Freed by task 270:
 kasan_save_stack+0x28/0x60
 kasan_set_track+0x28/0x40
 kasan_set_free_info+0x28/0x50
 __kasan_slab_free+0xfc/0x190
 kasan_slab_free+0x14/0x20
 kfree+0xe4/0x6e0
 sp_area_free+0x188/0x2f0
 sp_area_drop+0xe0/0x120
 remove_vma+0x84/0xa4

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index ff3baf415842..68fe9173dd57 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1312,6 +1312,7 @@ static int mm_add_group_init(pid_t tgid, struct mm_struct **pmm)
 	return ret;
 }
 
+static void sp_area_put_locked(struct sp_area *spa);
 /**
  * mg_sp_group_add_task() - Add a process to an share group (sp_group).
  * @tgid: the tgid of the task to be added.
@@ -1340,7 +1341,7 @@ static int mm_add_group_init(pid_t tgid, struct mm_struct **pmm)
  */
 int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 {
-	struct rb_node *p;
+	struct rb_node *p, *n;
 	struct mm_struct *mm;
 	struct sp_group *spg;
 	int ret = 0;
@@ -1392,17 +1393,26 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 	 * the spa in the spg may have changed. The newly created spa has been mapped
 	 * into the new task and we could get an errno of -EEXIST. Just ignore it.
 	 */
-	down_read(&spg->rw_lock);
-	for (p = rb_first(&spg->spa_root); p; p = rb_next(p)) {
+	down_write(&spg->rw_lock);
+	for (p = rb_first(&spg->spa_root); p; p = n) {
+		n = rb_next(p);
 		spa = container_of(p, struct sp_area, spg_link);
 
+		if (!atomic_inc_not_zero(&spa->use_count)) {
+			pr_warn("be careful, add new task(%d) to an exiting group(%d)\n",
+					tgid, spg_id);
+			continue;
+		}
+
 		ret = sp_map_spa_to_mm(mm, spa, prot, NULL, "add_task");
 		if (ret) {
 			pr_warn("mmap old spa to new task failed, %d\n", ret);
 			ret = 0;
 		}
+
+		sp_area_put_locked(spa);
 	}
-	up_read(&spg->rw_lock);
+	up_write(&spg->rw_lock);
 	/* We put the mm_struct later to protect the mm from exiting while sp_mmap */
 	mmput(mm);
 
@@ -2079,6 +2089,10 @@ static int sp_alloc_populate(struct mm_struct *mm, struct sp_area *spa,
 static int sp_k2u_populate(struct mm_struct *mm, struct sp_area *spa);
 
 #define SP_SKIP_ERR 1
+/*
+ * The caller should increase the refcnt of the spa to prevent that we map
+ * a dead spa into a mm_struct.
+ */
 static int sp_map_spa_to_mm(struct mm_struct *mm, struct sp_area *spa,
 			    unsigned long prot, struct sp_alloc_context *ac,
 			    const char *str)
-- 
2.17.1

